### PR TITLE
Accept java.util.Collection in SQL interpolation

### LINKED ISSUE #469: SQL interpolation behaves unexpectedly when passed a Java collection
If you accidentally end up with a Java collection and forget to convert it to a Scala collection, you get very confusing behaviour. The collection is happily accepted, but not unpacked, and just used (presumably) as an opaque POJO:

    scala> import collection.JavaConverters._
    import collection.JavaConverters._

    scala> sql"""update foo set bar=1 where id in (${Set(1, 2, 3)})""".statement
    res5: String = update foo set bar=1 where id in (?, ?, ?)

    scala> sql"""update foo set bar=1 where id in (${Set(1, 2, 3).asJava})""".statement
    res6: String = update foo set bar=1 where id in (?)

Note that the Java collection (the second one) doesn't unpack the Java set instance. I presume this applies to other collection types. I'm not sure if this should accept a Java collection or throw an exception (or maybe print a warning), but the current behaviour of trying to directly interpolate the opaque Java object without iterating through it is likely to end up in confusion (we end up with Java collections by accident all the time, but usually type checking in compilation finds these. However in this case you just get painful debugging where you wonder what's wrong with your query).

### PR DESCRIPTION
**Pull request by tkawachi**
Fixes #469 

This PR changed to converts all Traversable and java.util.Collection to immutable List before building query and params.

### CODE REVIEW

File: scalikejdbc-core/src/main/scala/scalikejdbc/SQLInterpolationString.scala [hunk 1]
```diff
@@ -14,7 +16,18 @@ class SQLInterpolationString(val s: StringContext) extends AnyVal {
     SQL[A](syntax.value).bind(syntax.parameters: _*)
   }
 
-  def sqls(params: Any*): SQLSyntax = SQLSyntax(buildQuery(params), buildParams(params))
+  def sqls(params: Any*): SQLSyntax = {
+    // Convert mutable collections to immutable Lists here to avoid
+    // mutation from another thread, which might cause a mismatch of
+    // the number of placeholders ("?") and parameters.
+    val convertedParams = params.map {
```
Comment (seratch) [inline scalikejdbc-core/src/main/scala/scalikejdbc/SQLInterpolationString.scala]: How about `fixedParams` or `immutableParams`?

Comment (tkawachi) [inline scalikejdbc-core/src/main/scala/scalikejdbc/SQLInterpolationString.scala]: The converted result doesn't have mutable 'collections', but might contain mutable objects. so `immutable` would be misleading. I feel `fixedParams` is fine. will rename it.

**Comment by tkawachi** [comment]
renamed the variable. (amended)

**Comment by seratch** [comment]
Thanks. LGTM. :+1:

**Comment by gakuzzzz** [comment]
Does it break compatibility of nested traversable like as `Set[Set[Int]]` ?

**Comment by tkawachi** [comment]
Right. Does it assumed to support such a nested collection type?

**Comment by gakuzzzz** [comment]
example, `in` query and using array data type in DB.

**Comment by gakuzzzz** [comment]
Maybe, there are no problem.
However I have no assurance……

**Comment by seratch** [comment]
Accepting Scala/Java collection values as-is is not supported right now. Furthermore, `java.sql.Array` which represents array type in DB is a fully different type from them. 

If I understand correctly, nested collection type values have been unsupported util now.

**Comment by gakuzzzz** [comment]
I agree :+1:

**Pull request merged**